0.40.0-rc.1: panic: duplicate metrics collector registration attempted

### What's wrong?

Starting Grafana Agent `docker.io/grafana/agent:v0.40.0-rc.0` with the config below results into a panic. The config works fine at `v0.37.3`

### Steps to reproduce

Starting Grafana Agent `docker.io/grafana/agent:v0.40.0-rc.0` with the config below.


i guess its an combination of `remote.http` and `loki.write`. Remove `remote.http` resolves the issue.

The http endpoint returns a JSON:

```json
{"resourceId": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg/providers/Microsoft.Compute/virtualMachines/agent-test-imds-mock", "name": "agent-test-imds-mock", "tagsList": [{"name": "cdt-operations-tier", "value": "mock"}]}
```

### System information

Linux

### Software version

0.40.0-rc.1

### Configuration

```text
remote.http "imds" {
    url            = coalesce(env("OPSSTACK_AGENT_IMDS_ENDPOINT"), "http://169.254.169.254") + "/metadata/instance/compute?api-version=2021-02-01"
    headers = {
        "Metadata" = "true",
    }
    poll_frequency = "86400s"
    poll_timeout   = "1s"
}


logging {
    level  = "warn"
    format = "json"
    write_to = [loki.write.remote.receiver]
}

loki.write "remote" {
    external_labels = {
        "vmName" = json_decode(nonsensitive(remote.http.imds.content))["name"],
        "os"     = constants.os,
        "id"     = to_lower(json_decode(nonsensitive(remote.http.imds.content))["resourceId"]),
    }

    endpoint {
        name      = "opsstack"
        url       = env("OPSSTACK_AGENT_REMOTE_URL") + "/loki/api/v1/push"
        headers   = {
            "Host" = coalesce(env("OPSSTACK_AGENT_REMOTE_HOSTNAME"), replace(env("OPSSTACK_AGENT_REMOTE_URL"), "https://", "")),
            "X-OpsStack-Resource-ID" = json_decode(nonsensitive(remote.http.imds.content))["resourceId"],
            "X-OpsStack-Resource-OS" = constants.os,
        }
        tenant_id = "agent"

        basic_auth {
            username = env("OPSSTACK_AGENT_REMOTE_USERNAME")
            password = env("OPSSTACK_AGENT_REMOTE_PASSWORD")
        }
    }
}

// Local Debugging

// prometheus.scrape "local" {
//   targets    = [/* ... */]
//   forward_to = [module.http.all.exports.prometheus.receiver]
// }
```


### Logs

```text
panic: duplicate metrics collector registration attempted

goroutine 180 [running]:
github.com/prometheus/client_golang/prometheus.(*wrappingRegisterer).MustRegister(0x4002f974d0, {0x400300e740?, 0x1, 0xa?})
	/go/pkg/mod/github.com/prometheus/client_golang@v1.18.0/prometheus/wrap.go:104 +0x158
github.com/grafana/agent/component/common/loki/client/internal.NewMarkerMetrics({0x8284290, 0x4002f974d0})
	/src/agent/component/common/loki/client/internal/metrics.go:22 +0x1c4
github.com/grafana/agent/component/common/loki/client.NewManager(0x0?, {0x8233420, 0x4002f7b540}, {0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0}, ...)
	/src/agent/component/common/loki/client/manager.go:92 +0x84
github.com/grafana/agent/component/loki/write.(*Component).Update(0x4002904fc0, {0x6a5dc20?, 0x4002f7be50?})
	/src/agent/component/loki/write/write.go:198 +0x468
github.com/grafana/agent/pkg/flow/internal/controller.(*BuiltinComponentNode).evaluate(0x4002fa8000, 0x0?)
	/src/agent/pkg/flow/internal/controller/node_builtin_component.go:286 +0x220
github.com/grafana/agent/pkg/flow/internal/controller.(*BuiltinComponentNode).Evaluate(0x4002f7a6e0?, 0xfdcf91d68730?)
	/src/agent/pkg/flow/internal/controller/node_builtin_component.go:241 +0x20
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).concurrentEvalFn(0x4002ef8c00, {0xfdcf91d68730, 0x4002fa8000}, {0x82a83e8, 0x4002fb0420}, {0x8259860, 0x4002fb0360}, 0x40028e8330)
	/src/agent/pkg/flow/internal/controller/loader.go:747 +0x374
github.com/grafana/agent/pkg/flow/internal/controller.(*Loader).EvaluateDependants.func2()
	/src/agent/pkg/flow/internal/controller/loader.go:699 +0x3c
github.com/grafana/agent/pkg/flow/internal/worker.(*workQueue).emitNextTask.func1()
	/src/agent/pkg/flow/internal/worker/worker_pool.go:181 +0x6c
github.com/grafana/agent/pkg/flow/internal/worker.(*fixedWorkerPool).start.func1()
	/src/agent/pkg/flow/internal/worker/worker_pool.go:87 +0x68
created by github.com/grafana/agent/pkg/flow/internal/worker.(*fixedWorkerPool).start in goroutine 1
	/src/agent/pkg/flow/internal/worker/worker_pool.go:80 +0x2c
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,10 @@ internal API changes are not present.
 Main (unreleased)
 -----------------
 
+### Bugfixes
+
+- Fix an issue where changing the configuration of `loki.write` would cause a panic. (@rfratto)
+
 v0.40.0-rc.1 (2024-02-23)
 -------------------------
 
@@ -27,7 +31,7 @@ v0.40.0-rc.1 (2024-02-23)
 
 ### Features
 
-- Modules have been redesigned to split the import logic from the instantiation. 
+- Modules have been redesigned to split the import logic from the instantiation.
   You can now define custom components via the `declare` config block and import modules via `import.git`, `import.http`, `import.string`, `import.file`. (@wildum)
 
 - A new `discovery.process` component for discovering Linux OS processes on the current host. (@korniltsev)
@@ -40,7 +44,7 @@ v0.40.0-rc.1 (2024-02-23)
 - Expose track_timestamps_staleness on Prometheus scraping, to fix the issue where container metrics live for 5 minutes after the container disappears. (@ptodev)
 
 - Introduce the `remotecfg` service that enables loading configuration from a
-  remote endpoint. (@tpaschalis) 
+  remote endpoint. (@tpaschalis)
 
 - Add `otelcol.connector.host_info` component to gather usage metrics for cloud users. (@rlankfo, @jcreixell)
 
@@ -102,7 +106,7 @@ v0.40.0-rc.1 (2024-02-23)
 - Fix an issue where agent logs are emitted before the logging format
   is correctly determined. (@hainenber)
 
-- Fix divide-by-zero issue when sharding targets. (@hainenber) 
+- Fix divide-by-zero issue when sharding targets. (@hainenber)
 
 - Fix bug where custom headers were not actually being set in loki client. (@captncraig)
 

--- a/component/common/loki/client/internal/metrics.go
+++ b/component/common/loki/client/internal/metrics.go
@@ -1,6 +1,9 @@
 package internal
 
-import "github.com/prometheus/client_golang/prometheus"
+import (
+	"github.com/grafana/agent/pkg/util"
+	"github.com/prometheus/client_golang/prometheus"
+)
 
 type MarkerMetrics struct {
 	lastMarkedSegment *prometheus.GaugeVec
@@ -19,7 +22,7 @@ func NewMarkerMetrics(reg prometheus.Registerer) *MarkerMetrics {
 		),
 	}
 	if reg != nil {
-		reg.MustRegister(m.lastMarkedSegment)
+		m.lastMarkedSegment = util.MustRegisterOrGet(reg, m.lastMarkedSegment).(*prometheus.GaugeVec)
 	}
 	return m
 }

--- a/component/common/loki/client/manager_test.go
+++ b/component/common/loki/client/manager_test.go
@@ -36,6 +36,28 @@ var (
 	metrics    = NewMetrics(prometheus.DefaultRegisterer)
 )
 
+// TestManager_NoDuplicateMetricsPanic ensures that creating two managers does
+// not lead to duplicate metrics registration.
+func TestManager_NoDuplicateMetricsPanic(t *testing.T) {
+	var (
+		host, _ = url.Parse("http://localhost:3100")
+
+		reg     = prometheus.NewRegistry()
+		metrics = NewMetrics(reg)
+	)
+
+	require.NotPanics(t, func() {
+		for i := 0; i < 2; i++ {
+			_, err := NewManager(metrics, log.NewLogfmtLogger(os.Stdout), testLimitsConfig, reg, wal.Config{
+				WatchConfig: wal.DefaultWatchConfig,
+			}, NilNotifier, Config{
+				URL: flagext.URLValue{URL: host},
+			})
+			require.NoError(t, err)
+		}
+	})
+}
+
 func TestManager_ErrorCreatingWhenNoClientConfigsProvided(t *testing.T) {
 	for _, walEnabled := range []bool{true, false} {
 		t.Run(fmt.Sprintf("wal-enabled = %t", walEnabled), func(t *testing.T) {

